qa/paqa_devs.c

```diff
@@ -52,83 +52,165 @@
+    0.020,
+    DEFAULT_FRAMES_PER_BUFFER,
+    0, /* numInputChannels */
+    1, /* numOutputChannels */
+    MODE_OUTPUT,
+    1, /* useCallback */
+    0, /* useNonInterleaved */
+};
+
+/* Runtime data used during the test. */
+typedef struct PaQaData
+{
+    const PaQaTestParameters *parameters;
+    // Dynamic state.
+    int              bytesPerSample;
+    volatile unsigned long    frameCounter;
+    volatile unsigned long    framesLeft;
+    unsigned long    framesPerBurst;
+    unsigned long    minFramesPerBuffer;
+    unsigned long    maxFramesPerBuffer;
+    unsigned long    framesDuration;
+    PaSineOscillator sineOscillators[MAX_TEST_CHANNELS];
+    void            *audioBuffer;
+} PaQaData;
 
 /****************************************** Prototypes ***********/
-static void TestDevices( int mode, int allDevices );
-static void TestFormats( int mode, PaDeviceIndex deviceID, double sampleRate,
-                         int numChannels );
-static int TestAdvance( int mode, PaDeviceIndex deviceID, double sampleRate,
-                        int numChannels, PaSampleFormat format );
+static int TestSingleStreamParameters(PaQaTestParameters parameters);
 static int QaCallback( const void *inputBuffer, void *outputBuffer,
                        unsigned long framesPerBuffer,
                        const PaStreamCallbackTimeInfo* timeInfo,
                        PaStreamCallbackFlags statusFlags,
                        void *userData );
 
+static void PaQaSetupData(PaQaData *myData,
+                          const PaQaTestParameters *parameters)
+{
+    myData->parameters = parameters;
+    myData->frameCounter = 0;
+    myData->framesLeft = (unsigned long) (parameters->sampleRate * parameters->durationSeconds);
+
+    myData->minFramesPerBuffer = UINT32_MAX;
+    myData->maxFramesPerBuffer = 0;
+
+    for (int channelIndex = 0; channelIndex < MAX_TEST_CHANNELS; channelIndex++)
+    {
+        myData->sineOscillators[channelIndex].phase = 0.0f;
+        myData->sineOscillators[channelIndex].phaseIncrement =
+        (2.0 * M_PI * LOWEST_FREQUENCY / parameters->sampleRate);
+    }
+
+    switch( parameters->format )
+    {
+        case paFloat32:
+        case paInt32:
+        case paInt24:
+            myData->bytesPerSample = 4;
+            break;
+            /*  case paPackedInt24:
+             myData->bytesPerSample = 3;
+             break; */
+        default:
+            myData->bytesPerSample = 2;
+            break;
+    }
+    myData->framesPerBurst = (parameters->framesPerBuffer == 0) ? 128 : parameters->framesPerBuffer;
+    if (parameters->useCallback == 0) {
+        /* We need our own buffer for blocking IO. */
+        int numChannels = (parameters->mode == MODE_OUTPUT)
+                ? parameters->numOutputChannels
+                : parameters->numInputChannels;
+        myData->audioBuffer = malloc(myData->bytesPerSample * numChannels * myData->framesPerBurst);
+    }
+}
+// TODO teardown delete audioBuffer
```

Comment:
I notice this is not done. I see why you haven't done it.

Reply:
Did it.